rest-api/tools/csv_importer.py

```diff
@@ -0,0 +1,96 @@
+"""Imports entities into the database based on a CSV file.
+
+Subclasses indicate in the constructor the name of the entity (for logging purposes),
+the DAO used to save the entity, the name of the primary key database ID field,
+the name of the external ID (referenced in the CSV file), and columns that must be populated
+with some value in the CSV.
+
+They then define _entity_from_row(row) to parse an entity out of a row dictionary.
+
+When dry_run flag is true, entities are not updated; instead logging indicates what would be
+updated.
+"""
+import csv
+import logging
+
+class CsvImporter(object):
```

Comment:
Consider using https://docs.python.org/2/library/abc.html#abc.ABCMeta if you want this to be an abstract base class.

Reply:
Interesting. We haven't been using this thus far for abstract classes. I don't necessarily mind introducing it... what would you say are the advantages over this approach?

Reply:
This was based on my reading of the below which seemed to imply you had abstract methods. Looks like you don't so this isn't needed. The only reason I generally use this (when applicable) is to ensure instantiation failure when subclassing without overriding abstract methods.

Reply:
There is one: subclasses must override _entity_from_row(self, row) in order for this to work.

That said, if you don't do it, the code will throw a useful error message the first time you call run() on a CSV file, which seems pretty much as good as failing on instantiation in this context.